idgenerator/idgenerator.go

```diff
@@ -0,0 +1,107 @@
+/*
+Package idgenerator contains several Span and Trace ID generators which can be
+used by the Zipkin tracer. Additional third party generators can be plugged in
+if they adhere to the IDGenerator interface.
+*/
+package idgenerator
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/openzipkin/zipkin-go/model"
+)
+
+var (
+	seededIDGen = rand.New(rand.NewSource(time.Now().UnixNano()))
+	// NewSource returns a new pseudo-random Source seeded with the given value.
+	// Unlike the default Source used by top-level functions, this source is not
+	// safe for concurrent use by multiple goroutines. Hence the need for a mutex.
+	seededIDLock sync.Mutex
+)
+
+// IDGenerator interface can be used to provide the Zipkin Tracer with custom
+// implementations to generate Span and Trace IDs.
+type IDGenerator interface {
+	SpanID() model.ID       // Generates a new Span ID
+	TraceID() model.TraceID // Generates a new Trace ID
+}
+
+// NewRandom64 returns an ID Generator which can generate 64 bit trace and span
+// id's
+func NewRandom64() IDGenerator {
+	return &randomID64{}
+}
+
+// NewRandom128 returns an ID Generator which can generate 128 bit trace and 64
+// bit span id's
+func NewRandom128() IDGenerator {
+	return &randomID128{}
+}
+
+// NewRandomTimestamped generates 128 bit time sortable traceid's and 64 bit
```

Comment:
the only use case I've seen so far unless there have been issues here, is to encode epoch seconds in a 128-bit trace ID. This is helpful for interop with AWS. If we don't want to do this by default in the go impl, probably having this as a pre-baked option makes sense. https://github.com/openzipkin/openzipkin.github.io/issues/49

Personally, I'd love to have no options except whether or not we use 128-bit trace IDs and only one option (the above one) for 128-bit trace IDs. Waiting until users ask for something is a nice engagement tool (correct me if I'm wrong and all the options here were asked for)

Reply:
It's an option I can really use for my own platform. So yes it's asked for ;)